lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -5750,6 +5752,41 @@ GenericSignatureBuilder::finalize(SourceLoc loc,
       }
     }
   }
+  
+  // Emit a diagnostic if we recorded any constraints where the constraint
+  // type was not constrained to a protocol or class. Provide a fix-it if
+  // allowConcreteGenericParams is true.
+  if (!invalidConstraints.empty()) {
+    for (auto constraint : invalidConstraints) {
+      auto loc = constraint.sourceLoc;
+      auto subjectType = constraint.subjectType;
+      auto constraintType = constraint.constraintType;
+      
+      Diags.diagnose(loc, diag::requires_conformance_nonprotocol,
+                     subjectType, constraintType);
+      
+      auto getNameWithoutSelf = [&](std::string subjectTypeName) {
```

Comment:
It makes the diagnostics look better. So, instead of saying `use 'Self.Foo == String' ... `, it says `use 'Foo == String' ...` and more accurately reflects the code the resulting code.